### PR TITLE
Fix transposed convolution in CPU w/o MKLDNN.

### PR DESCRIPTION
**Pull request by apeforest**
## Description ##
transposed convolution operator in CPU w/o MKLDNN is not working properly when dilation is set. This is because the mshadow library function `unpack_patch2col` and `pack_col2patch` generate incorrect results with dilation parameter. This PR replaced these two functions with MXNet native function `im2col` and `col2im`

This PR fixs issue https://github.com/apache/incubator-mxnet/issues/11203

Passed the local test in the issue:

```
import mxnet as mx

data = mx.nd.array(((0,0,0),
                    (0,1,0),
                    (0,0,0)))
kernel = mx.nd.array(((1,2,3),
                      (4,5,6),
                      (7,8,9)))

data_batch = data.expand_dims(0).expand_dims(0)
weight = kernel.expand_dims(0).expand_dims(0)
# initialize and set weight
conv = mx.gluon.nn.Conv2DTranspose(in_channels=1, channels=1,
                                   kernel_size=(3,3), padding=(2,2),
                                   strides=(2,2), dilation=(2,2))
conv.initialize()
conv.weight.set_data(weight)

weight.attach_grad()
with mx.autograd.record():
    l = conv(data_batch)
    print('conv forward')
    print(l.asnumpy())
l.backward()
print('conv backward')
print(weight.grad)
```
### Essentials ###
Please feel free to remove inapplicable items for your PR.
- [X] The PR title starts with [MXNET-$JIRA_ID], where $JIRA_ID refers to the relevant [JIRA issue](https://issues.apache.org/jira/projects/MXNET/issues) created (except PRs with tiny changes)
- [X] Changes are complete (i.e. I finished coding on this PR)
- [X] All changes have test coverage:
- Unit tests are added for small changes to verify correctness (e.g. adding a new operator)
- Nightly tests are added for complicated/long-running ones (e.g. changing distributed kvstore)
- Build tests will be added for build configuration changes (e.g. adding a new build option with NCCL)
- [X] Code is well-documented: 
- For user-facing API changes, API doc string has been updated. 
- For new C++ functions in header files, their functionalities and arguments are documented. 
- For new examples, README.md is added to explain the what the example does, the source of the dataset, expected performance on test set and reference to the original paper if applicable
- Check the API doc at http://mxnet-ci-doc.s3-accelerate.dualstack.amazonaws.com/PR-$PR_ID/$BUILD_ID/index.html
- [X] To the my best knowledge, examples are either not affected by this change, or have been fixed to be compatible with this change

### CODE REVIEW

**Comment by apeforest** [comment]
Just noticed that `test_deconvolution` in unit test has been disabled for a long time and was not re-enabled although the related issue https://github.com/apache/incubator-mxnet/issues/10973 is marked resolved. I will open another PR to re-enable this unit test first. Hence, marking this PR WIP due to lack of tests.

**Comment by apeforest** [comment]
@zhreshold @thomelane Please help to review. Thanks!

**Comment by zhreshold** [comment]
Can you verify the result in unittest?

File: src/operator/nn/deconvolution-inl.h [hunk 1]
```diff
@@ -226,22 +227,24 @@ class DeconvolutionOp {
     CHECK_EQ(in_data.size(), expected);
     CHECK_EQ(out_data.size(), 1U);
     Stream<xpu> *s = ctx.get_stream<xpu>();
+#if defined(__CUDACC__)
+    CHECK_EQ(s->blas_handle_ownership_, Stream<xpu>::OwnHandle)
+        << "Must init CuBLAS handle in stream";
```
Comment (pengzhao-intel) [inline src/operator/nn/deconvolution-inl.h]: "cuBLAS" is the official abbreviation :)

File: tests/python/unittest/test_gluon.py [hunk 2]
```diff
@@ -503,6 +503,40 @@ def test_deconv():
     # layer = nn.Conv3DTranspose(16, (3, 3, 3), layout='NDHWC', in_channels=4)
     # # check_layer_forward(layer, (1, 10, 10, 10, 4))
 
+@with_seed()
+def test_deconv_dilation():
```
Comment (zhreshold) [inline tests/python/unittest/test_gluon.py]: Since deconv is a really important OP,  I suggest to visit the original deconv test cases and add dilation > 1 cases alongside the old tests. This ensures better coverage than this single test case. 
Feel free to keep this unittest which LGTM as well.

File: src/operator/nn/deconvolution-inl.h [hunk 3]
```diff
@@ -251,76 +254,57 @@ class DeconvolutionOp {
...
         in_data[deconv::kWeight].get_with_shape<xpu, 3, DType>(wmat_shape, s);
-#if defined(__CUDACC__)
-    CHECK_EQ(s->blas_handle_ownership_, Stream<xpu>::OwnHandle)
-        << "Must init CuBLAS handle in stream";
-#endif
     const index_t nbatch = data.size(0);
     Tensor<xpu, 1, DType> workspace =
         ctx.requested[deconv::kTempSpace].get_space_typed<xpu, 1, DType>(
             Shape1(this->InitTemp(out.shape_, data.shape_)), s);
-    for (index_t i = 0; i < nbatch; i += nstep_) {
```
Comment (access2rohit) [inline src/operator/nn/deconvolution-inl.h]: Do you know what was "nstep_" doing earlier? It would help understand the problem with the earlier code.

File: src/operator/nn/deconvolution-inl.h [hunk 4]
```diff
@@ -373,55 +357,46 @@ class DeconvolutionOp {
...
-    CHECK_EQ(s->blas_handle_ownership_, Stream<xpu>::OwnHandle)
-        << "Must init CuBLAS handle in stream";
-#endif
 
     const index_t nbatch = data.size(0);
     Tensor<xpu, 1, DType> workspace =
         ctx.requested[deconv::kTempSpace].get_space_typed<xpu, 1, DType>(
             Shape1(this->InitTemp(grad.shape_, data.shape_)), s);
-    for (index_t i = 0; i < nbatch; i += nstep_) {
-      const index_t step = std::min(nstep_, nbatch - i);
```
Comment (access2rohit) [inline src/operator/nn/deconvolution-inl.h]: Again can you tell what was the purpose of "step" in the previous code?

**Comment by anirudhacharya** [comment]
@apeforest can you please rebase and resolve the merge conflicts?

File: src/operator/nn/deconvolution-inl.h [hunk 5]
```diff
@@ -485,7 +455,6 @@ class DeconvolutionOp {
   DeconvolutionParam param_;
   mshadow::Shape<2> shape_colunit_;
   mshadow::Shape<3> shape_dstunit_;
-  index_t nstep_;
```
[the remark this answers is not in the document]
Comment (apeforest) [reply src/operator/nn/deconvolution-inl.h]: The new im2col library does not support such step.

File: src/operator/nn/deconvolution-inl.h (hunk 3 above)
Comment (apeforest) [reply src/operator/nn/deconvolution-inl.h]: The new col2im library does not support such step.

File: src/operator/nn/deconvolution-inl.h (hunk 4 above)
Comment (apeforest) [reply src/operator/nn/deconvolution-inl.h]: I think it's used to convert multiple batch of image data into columns in the prevous library. However, it is not supported in the `col2im` method.

**Comment by abhinavs95** [comment]
@apeforest Could you please have a look at the CI failures?

**Comment by piyushghai** [comment]
@apeforest Gentle ping...

**Comment by Roshrini** [comment]
@apeforest Can you take a look at failing CI build?

**Comment by karan6181** [comment]
@apeforest Could you please provide an updates on this PR about your progress and thoughts so that the other community members get help from this. Thanks!

**Comment by apeforest** [comment]
@karan6181 The new function im2col has different signature and calling sequence from old col_unpack(). The changes fail in a few unit tests and I ended up re-implementing the operator itself. Given that current MKLDNN is default in CPU and it has no issue with Conv2DTranspose operator, I would like to treat this issue as lower priority and get it complete in a few weeks.

**Comment by piyushghai** [comment]
@mxnet-label-bot Update[pr-work-in-progress]

@apeforest Can you look into the CI failures ?

**Comment by wkcn** [comment]
@apeforest Hi! Any update in this PR? The PR is important: )

**Comment by PawelGlomski-Intel** [comment]
@szha with #11203 fixed by #20292, I believe this PR can be closed
